docs/puzzles/day7.md

```diff
@@ -2,12 +2,327 @@
 sidebar_position: 7
 ---
 
-# Day 7: Custom Customs
+import Solver from "../../../../website/src/components/Solver.js"
+
+# Day 7: The Treachery of Whales
+
+by [@tgodzik](https://github.com/tgodzik)
 
 ## Puzzle description
 
 https://adventofcode.com/2021/day/7
 
+## Part 1: Fast and craby
+
+The whale is trying to swallow out submarine! Fortunately, there is a small
+armada of crabs, let's call it a crabmada, in their tiny crabmarines that want
+to help us out! However, the crabs are not particularily efficient and will need
+our help, we need to align all their little crabmarines using the least fuel
+possible, so that they can blow a whole in the ocean floor.
+
+Firstly, let's start with modeling a single crabmarine:
+
+```scala
+case class Crabmarine(horizontal: Int):
+  def moveForward(): Crabmarine = this.copy(horizontal = horizontal + 1)
+  def moveBackward(): Crabmarine = this.copy(horizontal = horizontal - 1)
+```
+
+The crabmarine can move in two directions, forward and backwards.
+
+Let's translate our input into a list of `Crabmarine`:
+
+```scala
+val horizontalPositions = input.split(",").map(_.toInt).toList
+val crabmarines = horizontalPositions.map(horizontal => Crabmarine(horizontal))
+```
+
+A list of crabmarines can form a crabmada:
+
+```scala
+case class Crabmada(crabmarines: List[Crabmarine]):
+  // we don't want an empty list here!
+  require(crabmarines.nonEmpty)
+
+  def align() = ???
+```
+
+and it can have a method for aliging all its crabs.
+
+Now the tricky part, how can we figure out the horizontal position that
+minimizes the amount of fuel the crabs will use? We have only two possible
+moves, forwards and backwards, and we know for sure that all the edge crabs will
+need to move. Let's model out solution as doing one of two moves:
+
+- moving all the edge crabmarines with the current maximum horizontal position
+  backwards
+- moving all the edge crabmarines with the current minimal horizontal position
+  forwards
+
+In each interation we can move the crabmarines that use less fuel, which
+guarantees us that the result will be minimal at the end when all the crabs
+reach the same horizontal position.
+
+This can take a form of a recursive function:
+
+```scala
+  @tailrec
+  final def align(
+      situation: List[Crabmarine] = crabmarines,
+      fuelCost: Int = 0
+  ): Int =
+    val allTheSame = situation.forall(_.horizontal == situation.head.horizontal)
+    if allTheSame then fuelCost
+    else
+      val maxHorizontal = situation.maxBy(_.horizontal)
+      val minHorizontal = situation.minBy(_.horizontal)
+
+      val fuelCostForMax = situation.count {
+         crabmarine => crabmarine.horizontal == maxHorizontal.horizontal
+      }
+      val fuelCostForMin = situation.collect {
+        crabmarine => crabmarine.horizontal == minHorizontal.horizontal
+      }
+      if fuelCostForMax < fuelCostForMin then
```

Comment:
This does not typecheck.

Reply:
My bad, I was converting things for back for the part1. Will fix

Reply:
Fixed!